### PR TITLE
project addin

### PR DESCRIPTION
**Pull request by farooq222**

### CODE REVIEW

File: MyAddin/Form1.cs [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+        //this is the path where xml files are saved and loaded
+        //the input and output file will be found in AppData/roaming folder
+        string path = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
+        string fileName = Path.Combine(Environment.GetFolderPath(
+                Environment.SpecialFolder.ApplicationData), "input.xml");
+
+        string fileName1 = Path.Combine(Environment.GetFolderPath(
+                Environment.SpecialFolder.ApplicationData), "GenericFTS.xml");
+        String name1 = "";
```
Comment (Mdrbhatti) [inline MyAddin/Form1.cs:20]: In C#, `string` is an alias for `String`. Change it to `string` to remain consistent with rest of code.

File: MyAddin/Form1.cs [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+namespace MyAddin
+{
+    public partial class Form1 : Form
+    {
+        private XmlDocument docXML = new XmlDocument();
+
+        //this is the path where xml files are saved and loaded
+        //the input and output file will be found in AppData/roaming folder
+        string path = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
+        string fileName = Path.Combine(Environment.GetFolderPath(
```
Comment (Mdrbhatti) [inline MyAddin/Form1.cs:15]: Use more descriptive variable names. e.g. `inputFile`

File: MyAddin/Form1.cs [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+                Environment.SpecialFolder.ApplicationData), "input.xml");
+
+        string fileName1 = Path.Combine(Environment.GetFolderPath(
+                Environment.SpecialFolder.ApplicationData), "GenericFTS.xml");
+        String name1 = "";
+
+        // Xml tag for node, e.g. 'node' in case of <node></node>
+        public const string XmlNodeTag = "node";
```
Comment (Mdrbhatti) [inline MyAddin/Form1.cs]: Constant should be all CAPS!

File: MyAddin/Form1.cs [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+            try
+            {
+                treeView1.Nodes.Add("Failure");
+                treeView1.Nodes[0].Nodes.Add("Sporadic Timing Failure");
+                treeView1.Nodes[0].Nodes[0].Nodes.Add("Sporadic Duration Failure");
+                treeView1.Nodes[0].Nodes[0].Nodes[0].Nodes.Add("SporadicProlongation");
+                treeView1.Nodes[0].Nodes[0].Nodes[0].Nodes[0].Nodes.Add("Permanent Prolongation");
+                treeView1.Nodes[0].Nodes[0].Nodes[0].Nodes.Add("Permanent Duration Failure");
+                treeView1.Nodes[0].Nodes[0].Nodes[0].Nodes[1].Nodes.Add("Permanent Prolongation");
+                treeView1.Nodes[0].Nodes[0].Nodes[0].Nodes[1].Nodes.Add("Permament Interupt");
```
Comment (Mdrbhatti) [inline MyAddin/Form1.cs:64]: Use intermediate and descriptive variables to break this chain. E.g. `string nodeThirdChild = treeView1.Nodes[0].Nodes[0].....`
`nodeThirdChild.Add('ssss')....`

File: MyAddin/Form1.cs [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+            }
+            catch (Exception)
+            {
+                MessageBox.Show("Not Valid GenericFTS", "Error", MessageBoxButtons.OK);
```
Comment (Mdrbhatti) [inline MyAddin/Form1.cs:68]: Change `Not Valid` to `Invalid`

File: MyAddin/Form1.cs [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+        private void button1_Click(object sender, EventArgs e)
+        {
+            try
+            {
+                if (treeView1.SelectedNode == null)
+                {
+                    MessageBox.Show("Please Select a Node to add new node.", "Error", MessageBoxButtons.OK);
+                }
+
+                else if (name1 == "")
```
Comment (Mdrbhatti) [inline MyAddin/Form1.cs:98]: Better variable names

File: MyAddin/Form1.cs [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+                }
+                else
+                {
+                    this.Activate();
+                }
+            }
+            catch (Exception)
+            {
+                MessageBox.Show("Problem Closing the Application.", "Error", MessageBoxButtons.OK);
+                treeView1.Nodes.Clear();
```
Comment (Mdrbhatti) [inline MyAddin/Form1.cs:178]: If you have to write the same block  of code 2 or more times, move them to a new function.

File: MyAddin/Form1.cs [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+using System.Windows.Forms;
+using System.Xml;
+
+namespace MyAddin
+{
+    public partial class Form1 : Form
+    {
+        private XmlDocument docXML = new XmlDocument();
+
+        //this is the path where xml files are saved and loaded
```
Comment (shumailxyz) [inline MyAddin/Form1.cs:12]: I'd suggest to use multi-line ` /* */ `comment format if its not an inline comment. Looks more professional and better.

File: MyAddin/Form1.cs [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+            {
+                // disabling re-drawing of treeview till all nodes are added
+                treeView.BeginUpdate();
+                reader = new XmlTextReader(fileName);
+
+                TreeNode parentNode = null;
+
+                while (reader.Read())
+                {
+                    if (reader.NodeType == XmlNodeType.Element)
```
Comment (Mdrbhatti) [inline MyAddin/Form1.cs:246]: Convert this block of `if..elif` to a `switch` block.

File: MyAddin/MyAddinClass.cs [hunk 10]
```diff
@@ -0,0 +1,78 @@
+﻿using System;
+
+namespace MyAddin
+{
+    public class MyAddinClass
+	{
+		// define menu constants
+		const string menuHeader = "-&GenericFTS";
```
Comment (Mdrbhatti) [inline MyAddin/MyAddinClass.cs:8]: Consts should begin with a capital letter.

File: MyAddin/MyAddinClass.cs [hunk 11]
```diff
@@ -0,0 +1,78 @@
+﻿using System;
```
Comment (Mdrbhatti) [inline MyAddin/MyAddinClass.cs:1]: Fix indentation of this whole file.

File: MyAddin/Form1.cs [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+
+        // Xml attributes for node e.g. <node text="Sporadic Timing Failure"></node>
+        public const string XmlNodeTextAtt = "text";
+
+        public Form1()
+        {
+            try
+            {
+                InitializeComponent();
+                //this will chk if xml exist and load that
```
Comment (shumailxyz) [inline MyAddin/Form1.cs:33]: spellings of `check`

File: MyAddin/Form1.cs [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+
+        }
+
+        public void LoadXml()
+        {
+            try
+            {
+                ReadXMLFile(treeView1, fileName);
+            }
+            catch (Exception)
```
Comment (shumailxyz) [inline MyAddin/Form1.cs:79]: make use of exception block and throw exception error message atleast on console.

File: MyAddin/MyAddinClass.cs [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+			//No special processing required.
+			return "a string";
+		}
+
+		/// Called when user Clicks Add-Ins Menu item from within EA.
+		/// Populates the Menu with our desired selections
+		public object EA_GetMenuItems(EA.Repository Repository, string Location, string MenuName)
+		{
+
+            switch (MenuName)
```
Comment (Mdrbhatti) [inline MyAddin/MyAddinClass.cs:22]: Change `switch` to `if..else` block. (Switch should be used when 3-4+ conditionals in effect)

File: MyAddin/Form1.cs [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+            {
+            }
+        }
+
+        private void treeView1_AfterSelect(object sender, TreeViewEventArgs e)
+        {
```
Comment (shumailxyz) [inline MyAddin/Form1.cs:85]: empty function?

File: MyAddin/Form1.cs [hunk 16]
```diff
@@ -0,0 +1,317 @@
...
+            catch (Exception)
+            {
+                MessageBox.Show("Please Select a Node to add new node.", "Error", MessageBoxButtons.OK);
+            }
+            treeView1.SelectedNode = null;
+            textBox1.Clear();
+            return;
+        }
+
+        //here user give name for new node
```
Comment (shumailxyz) [inline MyAddin/Form1.cs:117]: use multi-line comment format please

File: MyAddin/Form1.cs [hunk 17]
```diff
@@ -0,0 +1,317 @@
...
+        private void textBox1_TextChanged(object sender, EventArgs e)
+        {
+            try
+            {
+                name1 = textBox1.Text;  //get name from user for new node
```
Comment (shumailxyz) [inline MyAddin/Form1.cs:122]: better variable names for `name` and `textBox1`

File: MyAddin/MyAddinClass.cs [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+
+		/// returns true if a project is currently opened
+		bool IsProjectOpen(EA.Repository Repository)
+		{
+			try
+			{
+				EA.Collection c = Repository.Models;
+				return true;
+			}
+			catch
```
Comment (Mdrbhatti) [inline MyAddin/MyAddinClass.cs:38]: You are not catching any exception here.

File: MyAddin/MyAddinClass.cs [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+				return false;
+			}
+		}
+
+		/// Called when user makes a selection in the menu.
+		public void EA_MenuClick(EA.Repository Repository, string Location, string MenuName, string ItemName)
+		{
+            try
+            {
+                GenericFTS();
```
Comment (Mdrbhatti) [inline MyAddin/MyAddinClass.cs:49]: `GenericFTS` doesn't throw any exception, so it shouldn't be surrounded by a `try..catch` block!

File: MyAddin/MyAddinClass.cs [hunk 20]
```diff
@@ -0,0 +1,78 @@
+﻿using System;
+
+namespace MyAddin
+{
+    public class MyAddinClass
```
Comment (shumailxyz) [inline MyAddin/MyAddinClass.cs:5]: better class name and namespace please

File: MyAddin/MyAddinClass.cs [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+	{
+		// define menu constants
+		const string menuHeader = "-&GenericFTS";
+
+        /// Called Before EA starts to check Add-In Exist
+        public String EA_Connect(EA.Repository Repository)
+		{
+			//No special processing required.
+			return "a string";
```
Comment (shumailxyz) [inline MyAddin/MyAddinClass.cs:14]: doesnt really make any sense

File: MyAddin/MyAddinClass.cs (hunk 14 above)
Comment (shumailxyz) [inline MyAddin/MyAddinClass.cs:22]: switch/case for only 1 case? normally we use switch/case for multuple cases and a default case

**Review by Mdrbhatti — COMMENTED** [review]
In short, try to name variables so anyone can tell what they are used for on a glance.
Don't use `try..except` blocks when the code NEVER throws any exceptions.

File: MyAddin/MyAddinClass.cs [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+
+		/// Called when user makes a selection in the menu.
+		public void EA_MenuClick(EA.Repository Repository, string Location, string MenuName, string ItemName)
+		{
+            try
+            {
+                GenericFTS();
+            }
+            catch (Exception)
+            {
```
Comment (shumailxyz) [inline MyAddin/MyAddinClass.cs:52]: throw error or notify user at least

File: MyAddin/MyAddinClass.cs [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+            }
+            
+        }
+
+        // calling form created in separate class
+        public void GenericFTS()
+        {
+            try
+            {
+                Form1 objForm1 = new Form1();
```
Comment (shumailxyz) [inline MyAddin/MyAddinClass.cs:62]: rename form1

**Review by shumailxyz — CHANGES_REQUESTED** [review]

File: MyAddin/Form1.cs [hunk 24]
```diff
@@ -0,0 +1,317 @@
...
+        public void SetAttributeValue(TreeNode node, string propertyName, string value)
+        {
+            if (propertyName == XmlNodeTextAtt)
+            {
+                node.Text = value;
+            }
+        }
+    }
+}
+
```
Comment (Mdrbhatti) [inline MyAddin/Form1.cs:315]: Remove new lines

**Comment by farooq222** [comment]
Thank you @Mdrbhatti and @shumail92 for reviewing the code.

**Pull request merged**